### PR TITLE
✨ Extend ControlPlane/ MD / MP builtin to include metadata

### LINKED ISSUE #10836: Make MD metadata available within MachineDeployment Builtins
### What would you like to be added (User Story)?

As a runtime extension developer, I'd like to be able to generate bootstrap templates based on metadata set on the machine deployments.

### Detailed Description

Cluster topology allows setting additional metadata onto the MachineDeployment. This metadata can be used to add conditional logic when generating patches via runtime extensions. Making this metadata available in the MachineDeployment builtins would allow for easier generation of conditional logic when managing multiple machine deployments per cluster.

### Anything else you would like to add?

_No response_

### Label(s) to be applied

/kind feature
/area clusterclass

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
This patch extends the MachineDeployment and MachinePool builtins to include metadata from topology MachineDeployment and topology MachinePool respectively.

**Which issue(s) this PR fixes**:
Fixes #10836 

/area clusterclass

### CODE REVIEW

**Comment by srm09** [comment]
The original issue only talks about `MachineDeployment` builtins, but when implementing I realized that this could be a requirement on the `MachinePool` builtins as well. Hence included that in the PR as well.

File: exp/runtime/hooks/api/v1alpha1/topologymutation_variable_types.go [hunk 1]
```diff
@@ -16,7 +16,11 @@ limitations under the License.
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/topologymutation_variable_types.go:16]: Please see your previous PR for all the places where the new builtin variable requires changes: https://github.com/kubernetes-sigs/cluster-api/pull/10822 :)

File: exp/runtime/hooks/api/v1alpha1/topologymutation_variable_types.go [hunk 2]
```diff
@@ -116,6 +120,9 @@ type MachineDeploymentBuiltins struct {
 	// being orchestrated.
 	Version string `json:"version,omitempty"`
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/topologymutation_variable_types.go:121]: Let's please add the same also to ControlPlaneBuiltins

File: exp/runtime/hooks/api/v1alpha1/topologymutation_variable_types.go [hunk 3]
```diff
@@ -116,6 +120,9 @@ type MachineDeploymentBuiltins struct {
...
 
+	// Metadata is the additional metadata set on the MachineDeployment
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/topologymutation_variable_types.go:123]:
```suggestion
	// Metadata is the metadata set on the MachineDeployment
```
I think

File: internal/controllers/topology/cluster/patches/variables/variables.go [hunk 4]
```diff
@@ -182,6 +182,16 @@ func MachineDeployment(mdTopology *clusterv1.MachineDeploymentTopology, md *clus
 	if md.Spec.Replicas != nil {
 		builtin.MachineDeployment.Replicas = ptr.To[int64](int64(*md.Spec.Replicas))
 	}
+	if mdTopology.Metadata.Labels != nil || mdTopology.Metadata.Annotations != nil {
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/variables/variables.go:185]: Let's please take the labels & annotations from md not from mdTopology.

The ones from md contain *all* labels & annotations that we eventually set on the MachineDeployment

File: internal/controllers/topology/cluster/patches/variables/variables.go [hunk 5]
```diff
@@ -182,6 +182,16 @@ func MachineDeployment(mdTopology *clusterv1.MachineDeploymentTopology, md *clus
...
+		metadata := &clusterv1.ObjectMeta{}
+		if mdTopology.Metadata.Labels != nil {
+			metadata.Labels = mdTopology.Metadata.Labels
+		}
+		if mdTopology.Metadata.Annotations != nil {
+			metadata.Annotations = mdTopology.Metadata.Annotations
+		}
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/variables/variables.go:192]: I think the nil checking inside the if is not necessary

**Comment by sbueringer** [comment]
> The original issue only talks about `MachineDeployment` builtins, but when implementing I realized that this could be a requirement on the `MachinePool` builtins as well. Hence included that in the PR as well.

Sounds good. Let's do the same for control plane.

**Comment by srm09** [comment]
/retitle ✨ Extend ControlPlane/Deployment/Pool builtin to include metadata

**Comment by sbueringer** [comment]
Thank you!

/lgtm
/approve

**Pull request merged**